### PR TITLE
Allow publishing from workspace root.

### LINKED ISSUE #7345: Can't publish from workspace root, no -p option
Unlike other `cargo` commands, `cargo publish` doesn't accept `-p` or `--package` option, so `cargo publish` doesn't work at all at the workspace root:
```
$ cargo publish
error: manifest path `.../Cargo.toml` is a virtual manifest, but this command requires running against an actual package in this workspace
$ cargo publish -p name 
error: Found argument '-p' which wasn't expected, or isn't valid in this context
```

Lack of the `-p` option seems like an oversight.

### PR DESCRIPTION
**Pull request by tcmal**
Adds -p, --workspace, and --exclude to package and publish commands.

Uses ephemeral workspaces to avoid changing the existing functions too much.
There might be more `Finished dev [unoptimized + debuginfo] target` messages when packaging than there should be, I couldn't figure out what was generating them.
The tests aren't super extensive, as all the specs from arguments code should already be tested elsewhere.

Closes #7345

### CODE REVIEW

**Comment by ehuss** [comment]
Thanks for the PR!  I think for now, there should not be a `--workspace` flag.  There is an issue with uploading multiple packages in quick succession.  There is a brief period after a package is uploaded that it is not available in the index, which can cause published crates to fail if they require the new versions.  Additionally, when publishing multiple inter-related packages, the publish process will need to start from the leaves of the dependency tree and work its way upward.  For now, can this be kept to only one package?

**Comment by tcmal** [comment]
I'm not sure why it's failing, I'm going to rebase and see if that fixes it

File: src/cargo/ops/cargo_package.rs [hunk 1]
```diff
@@ -135,9 +137,46 @@ pub fn package(ws: &Workspace<'_>, opts: &PackageOpts<'_>) -> CargoResult<Option
...
+    opts: &PackageOpts<'_>,
+) -> CargoResult<Vec<Option<FileLock>>> {
+    let pkgs = ws_root.members_with_features(
+        &opts.to_package.to_package_id_specs(ws_root)?,
+        &opts.cli_features,
+    )?;
+    let mut dsts = Vec::with_capacity(pkgs.len());
+
+    for (pkg, cli_features) in pkgs {
+        let ws = Workspace::ephemeral(
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_package.rs:155]: Instead of creating an ephemeral workspace per-package, could this pass in the existing workspace and the `Package` itself perhaps that's being packaged?

File: src/cargo/ops/registry.rs [hunk 2]
```diff
@@ -50,16 +50,28 @@ pub struct PublishOpts<'cfg> {
...
-    let mut publish_registry = opts.registry.clone();
+pub fn publish(root_ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
+    let specs = opts.to_publish.to_package_id_specs(root_ws)?;
+    let mut pkgs = root_ws.members_with_features(&specs, &opts.cli_features)?;
+
+    anyhow::ensure!(pkgs.len() == 1, "can only publish one package at a time.");
+
+    let (pkg, cli_features) = pkgs.pop().unwrap();
 
+    let ws = Workspace::ephemeral(
```
Comment (alexcrichton) [inline src/cargo/ops/registry.rs:68]: Like above I think it would be best to avoid ephemeral workspaces where possible, so I think the original `ws` should be used here if possible.

File: src/cargo/ops/registry.rs [hunk 3]
```diff
@@ -50,16 +50,28 @@ pub struct PublishOpts<'cfg> {
...
 }
 
-pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
-    let pkg = ws.current()?;
-    let mut publish_registry = opts.registry.clone();
+pub fn publish(root_ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
+    let specs = opts.to_publish.to_package_id_specs(root_ws)?;
+    let mut pkgs = root_ws.members_with_features(&specs, &opts.cli_features)?;
+
+    anyhow::ensure!(pkgs.len() == 1, "can only publish one package at a time.");
```
Comment (alexcrichton) [inline src/cargo/ops/registry.rs:64]: Am I correct in reading this in that this will support many `-p` flags and and `--workspace`, but we'll just return an error if they get matched? If so, can this error be elaborated a bit as to why multiple packages aren't supported at this time? (perhaps linking to a tracking issue?)

Comment (tcmal) [reply src/cargo/ops/registry.rs:64]: It won't take --workspace, but it will take multiple -p flags. Like @ehuss said this is due to an issue with crates.io, I couldn't find any tracking issue on their repo.

File: src/bin/cargo/commands/publish.rs [hunk 4]
```diff
@@ -18,6 +18,7 @@ pub fn cli() -> App {
         ))
         .arg_target_triple("Build for the target triple")
         .arg_target_dir()
+        .arg_package_spec_simple("Package to publish")
```
Comment (ehuss) [inline src/bin/cargo/commands/publish.rs:21]: Would it be possible to use `.arg_package()` instead?  That way, it would validate that only one `-p` flag is used by the CLI parser.

File: src/cargo/ops/cargo_package.rs [hunk 5]
```diff
@@ -81,7 +78,7 @@ pub fn package(ws: &Workspace<'_>, opts: &PackageOpts<'_>) -> CargoResult<Option
     if !pkg.manifest().exclude().is_empty() && !pkg.manifest().include().is_empty() {
         config.shell().warn(
             "both package.include and package.exclude are specified; \
-             the exclude list will be ignored",
+            the exclude list will be ignored",
```
Comment (ehuss) [inline src/cargo/ops/cargo_package.rs:81]: There seems to be several unrelated whitespace changes, can you try to avoid the extraneous changes?

**Review by ehuss — COMMENTED** [review]
Sorry about the long delay.

**Comment by ehuss** [comment]
Thanks!

@bors r+

File: tests/testsuite/publish.rs [hunk 6]
```diff
@@ -1706,3 +1734,73 @@ Caused by:
...
+                workspace = ".."
+            "#,
+        )
+        .file("bar/src/main.rs", "fn main() {}")
+        .build();
+
+    p.cargo("publish --no-verify --token sekrit -p foo")
+        .with_stderr(&format!(
+            "\
+[UPDATING] `{reg}` index
```
Comment (ehuss) [inline tests/testsuite/publish.rs:1780]: The exacting text of the `Updating` line isn't important to check here, so I think something like `[UPDATING] [..]` should be fine (without the format!).

**Comment by ehuss** [comment]
Thanks!  I pushed a small change to remove the unnecessary `format!`.

@bors r+
